### PR TITLE
Date field revamp

### PR DESCRIPTION
**Pull request by jomarmontuya**
closes https://github.com/zesty-io/manager-ui/issues/2414

- [x] Check with andres if timestamp is needed
- [x] Better approach on getting the zesty-orange background
- [x] Deployment to stage for early VQA and bugs ask @theofficialnar

### CODE REVIEW

**Comment by jomarmontuya** [comment]
image

**Comment by jomarmontuya** [comment]
https://github.com/zesty-io/manager-ui/assets/61284357/b2cf9e7d-b445-4979-9ad8-647227a86e5e

@zcolah This is the initial draft preview of this component. I think it would be better if we can ask @theofficialnar  to deploy it for you to see it yourself.

**Comment by zcolah** [comment]
A few functionalities are not present: 
1. The Date Selection Pop Up should appear and update when typing
2. If the field is clear with no date prior to this, then open the date picker with the current date (aka today’s date) selected

**Comment by jomarmontuya** [comment]
@agalin920  @theofficialnar 

Hello both,  Seeking your help, on how we can possibly achieve this experience, in our calendar, I have been researching on how to make it possible using core mui-x and can't seem to find a way to copy the same experience.

I would assume we have to write custom logic and override mui-x datepicker functionality to make it work, but I wanted to hear your thoughts, maybe you've encounter the same issues in the past. 

I was able to make it work but only partials not as fluid as the video below, and the core mui-x only supports custom inputs  but only specifics to defined format.

https://github.com/zesty-io/manager-ui/assets/61284357/e01fd291-c891-4d5f-a618-6a8cb23b7cdb

**Comment by jomarmontuya** [comment]
@agalin920  Thanks Andres, I will complete first the other requirements and get back to you on this one.

**Comment by jomarmontuya** [comment]
@agalin920 

```
 1. The Date Selection Pop Up should appear and update when typing
```
I was able to make the popup appear and set back the focus back to textfield when the TextField  is click, however this component needs to have state if we need realtime updates to the calendar picker. Because currently the picker date only updates after I complete typing the date in MMM DD, YYYY format.

```
2. If the field is clear with no date prior to this, then open the date picker with the current date (aka today’s date) selected
```

@zcolah  can you confirm this, as this is already implemented during our previous review. or I'm missing something?  

https://github.com/zesty-io/manager-ui/assets/61284357/04723027-d90b-4805-a52c-e99a2cb57af9


```
3. The experience of typing in a date is not fluid as per the expectations set. Please see requests and Google Calendar as a behavior reference.
```
This relates to the issue number one, also it looks like based on the requirements, we need to be able to accept different format inputs in the textfield. This is not possible at the moment with mui-x datePicker as we have to pick one format that we can use. 

image

```
 4. Check the padding of the icon button
```
This will get solved on the inputAdornment theme change is merged.

**Comment by zcolah** [comment]
@jomarmontuya I just met with @agalin920 and here are our two key takeaways: 

1. If the complexity to implement date formats 6, 7, 8 is too high and will take a full week, please do not do it. We will do it for a future iteration. 

2. The Loom video below should provide more context on what this means: If the field is clear with no date prior to this, then open the date picker with the current date (aka today’s date) selected
 
https://www.loom.com/share/4c2c18a4e60b4ec3ba81292381e38f69?from_recorder=1&focus_title=1

**Comment by jomarmontuya** [comment]
@zcolah 

https://www.loom.com/share/3825bf5e763a49e881687702ddca7012

here's the updated one please let me know what you think about this.

**Comment by zcolah** [comment]
Hi @jomarmontuya 
Thank you for this update. Currently it does fill the date in as expected but notice how it has the year part highlighted/selected. Could we instead have the month part be highlighted? 

CleanShot 2024-03-06 at 18 39 29

**Comment by jomarmontuya** [comment]
@zcolah  Thank you! here's the updated one 

https://www.loom.com/share/89c91db9a2b048cda5f22dcc105760b1

**Comment by jomarmontuya** [comment]
@zcolah 

if there's existing value in the dateField saved in the database, then the user click the clear the datefield will be null, but when they open the picker again what should be the default selected date?

- [ ] Current saved date from database?
- [ ] Current day similar to as if empty?

cc: @agalin920

File: src/apps/content-editor/src/app/components/Editor/Field/Field.tsx [hunk 1]
```diff
@@ -905,28 +905,21 @@ export const Field = ({
...
-              onChange={(date) => onDateChange(date, name, datatype)}
-              error={errors && Object.values(errors)?.some((error) => !!error)}
-            />
-          </Box>
+          <FieldTypeDate
+            name={name}
+            required={required}
+            value={value ? new Date(value) : null}
+            format="MMM dd, yyyy"
+            onChange={(date) => onDateChange(date, name, datatype)}
```
[the remark this answers is not in the document]
Comment (jomarmontuya) [reply src/apps/content-editor/src/app/components/Editor/Field/Field.tsx:920]: @agalin920  It is emitting localize date

File: src/shell/components/FieldTypeDate/index.tsx [hunk 2]
```diff
@@ -1,30 +1,140 @@
...
+                    mt: 1,
+                  },
+                },
+                day: {
+                  /*
+                   * Override the default today's background color to match with the theme
+                   */
+                  sx: {
+                    "&.MuiPickersDay-today": {
+                      background: "#FF5D0A",
```
[the remark this answers is not in the document]
Comment (jomarmontuya) [reply src/shell/components/FieldTypeDate/index.tsx:95]: I'm working on update on the theme level as this date picker will be used as a component and can appear on different places

File: src/shell/components/FieldTypeDate/index.tsx [hunk 3]
```diff
@@ -1,30 +1,140 @@
...
+       */
+      if (triggerPickerRef.current) triggerPickerRef.current?.click();
+    };
+
+    const onDatePickerOpen = () => {
+      if (textFieldRef.current) {
+        /**
+         * Add timeout to ensure the datepicker is open before setting the value
+         */
+        setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (jomarmontuya) [reply src/shell/components/FieldTypeDate/index.tsx:49]: by default when browser see a diagog box, the focus will be set to the closes focusable element inside the dialog.

I've added a setTimeout to delay the focus on textfield, to ensure that the dialog is fully open and focus was set, before moving the focus to the textfield.

See more details here -> https://web.dev/learn/html/dialog#:~:text=By%20default%2C%20when%20a%20dialog,when%20the%20dialog%20is%20opened.

File: src/shell/components/FieldTypeDate/index.tsx [hunk 4]
```diff
@@ -15,16 +19,85 @@ export const FieldTypeDate = ({
...
+              },
+              day: {
+                /*
+                 * Override the default today's background color to match with the theme
+                 */
+                sx: {
+                  "&.MuiPickersDay-today": {
+                    background: "#FF5D0A",
+                    border: "none",
+                    color: "white",
```
[the remark this answers is not in the document]
Comment (zcolah) [reply src/shell/components/FieldTypeDate/index.tsx]: This should be primary.main

[the remark this answers is not in the document]
Comment (zcolah) [reply src/shell/components/FieldTypeDate/index.tsx]: You can check this: https://zesty-io.github.io/material/?path=/story/theme--default

CleanShot 2024-03-11 at 20 32 22

**Comment by jomarmontuya** [comment]
@shrunyan  - >https://github.com/zesty-io/material/pull/93

Please see related PR to move this for VQA

**Comment by jomarmontuya** [comment]
@zcolah  This is up on dev you may proceed with VQA

**Review by zcolah — CHANGES_REQUESTED** [review]
@jomarmontuya

See Loom video that describes all of the issues below in more detail: https://www.loom.com/share/83204fcc82724beda555a7e49429f2e5

 1. Reduce the lag in showing the current date when the user opens the modal for the first time
 2. If possible please add a cursor indicator so that the user knows they are typing and knows where they are typing
 3. Add ability to update the entire date when it is highlighted/selected
 4. Once a user has selected a date via the Calendar pop up, the date input can go back to it's default non selected state. It does not need to be in active state and the year does not need to be highlighted.
 5. If there is an invalid date, then please immediately display an error message to a user. [See Figma design.](https://www.figma.com/file/zOI7oSH3fG1XDmmPzsP6PX/Zesty-3.0-Revamp?type=design&node-id=4080%3A27763&mode=dev&t=sOKSk2D0KnWplKAw-1)

**Comment by jomarmontuya** [comment]
@zcolah with this feedback I've got I think I had to do a completely different approach here instead for the textfield.

Instead of using the mui datepicker field out of the box I might have to use a custom textfield to allow user a full typing experience, because as of now I don't think there's a way to add cursor to the textfield, as it would only allow you to edit the date base on the format we assign which is "MMM DD, YYYY" that's why you can only change it as a whole not individual characters

@agalin920 I was thinking the solution we talked about last time, using a custom textfield with regex to make it work.

**Comment by jomarmontuya** [comment]
https://www.loom.com/share/138bafd09fc84528831118b562a96e73?sid=3433482c-0755-4ceb-9ba7-c61bde0ca290

@zcolah  here's the update regarding the datefield revamp 

cc @agalin920

**Comment by zcolah** [comment]
Looks good, will test

**Pull request merged**